docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+most sitatuations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero sitatuation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the cloud provider.
+
+An optional field is proposed for users to add to their Infrastructure Machine Templates which will
+carry information about the CPU, memory, and GPU requirements for the machines created from the template.
+This field will then be copied to the status of any MachineSet or MachineDeployment with an infrastructure
+reference to the template during reconciliation. The cluster autoscaler will then read this information
+from the status field to perform its calculations.
+
+A user may override the field in the associated infrastructure template by applying annotations to the
+MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+By keeping this information with the MachineSet and MachineDeployment resource objects, we reduce the
+need for the cluster autoscaler to have a deep knowledge of Cluster API internals.
+
+### User Stories
+
+- Detail the things that people will be able to do if this proposal is implemented.
+- Include as much detail as possible so that people can understand the "how" of the system.
+- The goal here is to make this feel real for users without getting bogged down.
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
+cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
+a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.
+
+#### Story 2
+
+As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
+without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
+not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
+creation of nodes that will not consume resources until they are required by applications on my cluster.
+
+### Implementation Details/Notes/Constraints
+
+A new common API type will be created named `ResourceHints`, its definition will look like this:
+
+```
+type ResourceHints struct {
+    vCPU   *string `json:"vCPU,omitempty"`
+    Memory *string `json:"memory,omitempty"`
+    GPU    *string `json:"gpu,omitempty"`
+}
+```
+
+This new API type can be used with a machine template to inform about the resources used
+by individual machines in the group. For example, here is a `DockerMachineTemplate` with
+the integrated `ResourceHints`
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  resourceHints:
```

Comment:
Wondering if this field name should somehow link explicitly to the autoscaler, in order to avoid potential confusion in the user that might assume those field could be used to override the machine type selection or as way to customise default machine sizes.